File: frontend/src/services/balanceador.ts
// Serviço do módulo "Balanceador de Agenda".
// MOCK: leitura real (diagnóstico/matriz/detalhe do pool); escrita simulada.

import { apiFetch } from "@/lib/api-client";

const BASE = "/api/v1/balanceador";

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let detail = `Erro ${res.status}`;
    try {
      detail = (await res.json())?.detail || detail;
    } catch {
      /* ignore */
    }
    throw new Error(detail);
  }
  return res.json() as Promise<T>;
}

export type Situacao = "atrasado" | "fatal_hoje" | "futuro" | "sem_prazo";

export interface Colaborador {
  id: number;
  nome: string;
  cargo: string | null;
  is_supervisor: boolean;
  atrasado: number;
  fatal_hoje: number;
  futuro: number;
  sem_prazo: number;
  total: number;
}

export interface MatrizItem {
  pessoa_id: number;
  subtipo: string;
  total: number;
  atrasado: number;
  fatal_hoje: number;
}

export interface TarefaDetalhe {
  l1_task_id: number | null;
  subtipo: string | null;
  descricao?: string | null; // assunto/anotações (vem no live; no snapshot é enriquecido à parte)
  cnj?: string | null;
  pasta?: string | null;
  uf?: string | null;
  prazo: string | null;
  situacao: Situacao;
}

// ── LIVE: pendentes de uma pessoa direto do L1 (matriz + detalhe) ──
export interface LivePessoaSub {
  subtipo: string;
  total: number;
  atrasado: number;
  fatal_hoje: number;
}
export interface LivePessoa {
  pessoa_id: number;
  nome: string | null;
  resolvido: boolean;
  total_real?: number | null; // total de pendentes COM prazo no L1 (pode ser > carregadas)
  carregadas?: number; // quantas vieram (teto das mais urgentes)
  capado?: boolean; // true = estourou o teto; há mais além das mais urgentes
  subtipos: LivePessoaSub[];
  tarefas: TarefaDetalhe[];
}
export async function getLivePessoa(team: string, pessoaId: number, dias: number): Promise<LivePessoa> {
  const qs = new URLSearchParams({ team, pessoa_id: String(pessoaId), dias: String(dias) });
  return json(await apiFetch(`${BASE}/live-pessoa?${qs.toString()}`));
}

export async function getDiagnostico(team: string): Promise<Colaborador[]> {
  const r = await json<{ colaboradores: Colaborador[] }>(await apiFetch(`${BASE}/diagnostico?team=${team}`));
  return r.colaboradores;
}

export async function getMatriz(team: string, pessoaIds: number[], dias: number): Promise<MatrizItem[]> {
  const qs = new URLSearchParams({ team, pessoas: pessoaIds.join(","), dias: String(dias) });
  const r = await json<{ matriz: MatrizItem[] }>(await apiFetch(`${BASE}/redistribuir?${qs.toString()}`));
  return r.matriz;
}

export async function getTarefas(
  team: string,
  pessoaId: number,
  subtipo: string,
  dias: number,
): Promise<TarefaDetalhe[]> {
  const qs = new URLSearchParams({ team, pessoa_id: String(pessoaId), subtipo, dias: String(dias) });
  const r = await json<{ tarefas: TarefaDetalhe[] }>(await apiFetch(`${BASE}/tarefas?${qs.toString()}`));
  return r.tarefas;
}

// Descrição (assunto/anotações) ao vivo do L1 — não vem no snapshot.
export async function getDescricoes(team: string, ids: number[]): Promise<Record<number, string | null>> {
  if (!ids.length) return {};
  const qs = new URLSearchParams({ team, ids: ids.join(",") });
  const r = await json<{ descricoes: Record<number, string | null> }>(
    await apiFetch(`${BASE}/descricoes?${qs.toString()}`),
  );
  return r.descricoes;
}

// ── Modelo local de "mudanças pendentes" (escrita simulada no mock) ──
export interface MovePendente {
  id: string; // chave local
  fromId: number;
  fromNome: string;
  toId: number;
  toNome: string;
  subtipo: string;
  qtd: number;
  individual: boolean; // true = tarefas escolhidas a dedo; false = em massa por número
  taskIds?: number[];
}

// ── Log de redistribuição (aba Relatórios) ──
export interface RedistribuicaoLog {
  id: number;
  criado_em: string | null;
  criado_por_nome: string | null;
  total_movimentos: number;
  total_tarefas: number;
  origem: string;
  detalhe: MovePendente[];
}

export async function registrarLog(team: string, movimentos: MovePendente[]): Promise<void> {
  const res = await apiFetch(`${BASE}/log?team=${team}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ movimentos }),
  });
  if (!res.ok) throw new Error(`Erro ${res.status} ao registrar o log`);
}

export async function listarLogs(team: string): Promise<RedistribuicaoLog[]> {
  const r = await json<{ logs: RedistribuicaoLog[] }>(await apiFetch(`${BASE}/logs?team=${team}`));
  return r.logs;
}

File: app/services/performance/balanceador.py
"""Balanceador de Agenda — leituras do pool de pendentes pro supervisor
diagnosticar carga e redistribuir entre colaboradores.

MOCK (2026-06-29): lê do snapshot `perf_l1_tarefa` (o mesmo do Minha Equipe).
Na versão real a fila vem AO VIVO do L1 e a escrita reatribui de fato
(API PATCH p/ normal, POST ModalEnvolvimentoEmLote p/ Workflow — já provado).
Aqui é tudo read-only.

Escopo redistribuível: pendentes, fora dos subtipos `Acompanhar*` (são segmento
de tarefa já iniciada — não se redistribui).
"""

import datetime as _dt

from sqlalchemy import text
from sqlalchemy.orm import Session

_BRT = "America/Sao_Paulo"
_HOJE = f"(now() AT TIME ZONE '{_BRT}')::date"
_PRAZO = f"(t.prazo_previsto AT TIME ZONE '{_BRT}')::date"
_NAO_ACOMP = "lower(coalesce(t.subtipo,'')) NOT LIKE 'acompanhar%'"

try:
    from zoneinfo import ZoneInfo

    _TZ = ZoneInfo(_BRT)
except Exception:  # pragma: no cover
    _TZ = None


def _hoje_brt() -> _dt.date:
    return (_dt.datetime.now(tz=_TZ) if _TZ else _dt.datetime.now()).date()


# Cache (vida do processo) do mapa nome_norm -> contact_id do L1. Usuários mudam
# raramente; evita bater get_all_users (~310, ~6s) a cada chamada live.
_USER_MAP: dict = {}


def _user_map() -> dict:
    if not _USER_MAP:
        from app.services.legal_one_client import LegalOneApiClient
        from app.services.performance.seed import norm

        for u in LegalOneApiClient().get_all_users():
            nome = u.get("name")
            if nome and u.get("id"):
                _USER_MAP[norm(nome)] = u["id"]
    return _USER_MAP


def _periodo_clause(dias: int) -> str:
    """Janela: atrasados + os que vencem nos próximos `dias` (+ sem prazo).
    dias<=0 = tudo (sem teto)."""
    if dias and dias > 0:
        return f" AND (t.prazo_previsto IS NULL OR {_PRAZO} <= {_HOJE} + :dias)"
    return ""


class BalanceadorService:
    def __init__(self, db: Session):
        self.db = db

    def diagnostico(self, team: str) -> list[dict]:
        """Por colaborador do time: pendentes atrasadas / fatais hoje / futuras."""
        rows = self.db.execute(
            text(
                f"""
                SELECT p.id, p.nome, p.cargo, p.is_supervisor,
                  count(t.id) FILTER (WHERE t.prazo_previsto IS NOT NULL AND {_PRAZO} < {_HOJE}) AS atrasado,
                  count(t.id) FILTER (WHERE {_PRAZO} = {_HOJE}) AS fatal_hoje,
                  count(t.id) FILTER (WHERE t.prazo_previsto IS NOT NULL AND {_PRAZO} > {_HOJE}) AS futuro,
                  count(t.id) FILTER (WHERE t.prazo_previsto IS NULL) AS sem_prazo,
                  count(t.id) AS total
                FROM perf_pessoa p
                LEFT JOIN perf_l1_tarefa t
                  ON t.pessoa_id = p.id AND t.status = 'Pendente' AND {_NAO_ACOMP}
                WHERE p.equipe = :team AND p.ativo
                GROUP BY p.id, p.nome, p.cargo, p.is_supervisor
                ORDER BY p.is_supervisor DESC, atrasado DESC, futuro DESC, p.nome
                """
            ),
            {"team": team},
        ).fetchall()
        return [
            {
                "id": r.id, "nome": r.nome, "cargo": r.cargo, "is_supervisor": r.is_supervisor,
                "atrasado": r.atrasado, "fatal_hoje": r.fatal_hoje, "futuro": r.futuro,
                "sem_prazo": r.sem_prazo, "total": r.total,
            }
            for r in rows
        ]

    def redistribuir_matriz(self, team: str, pessoa_ids: list, dias: int) -> list[dict]:
        """Subtipos × colaborador (contagens) pra os escolhidos, dentro do período."""
        if not pessoa_ids:
            return []
        rows = self.db.execute(
            text(
                f"""
                SELECT t.pessoa_id, coalesce(t.subtipo, '(sem subtipo)') AS subtipo,
                  count(*) AS total,
                  count(*) FILTER (WHERE t.prazo_previsto IS NOT NULL AND {_PRAZO} < {_HOJE}) AS atrasado,
                  count(*) FILTER (WHERE {_PRAZO} = {_HOJE}) AS fatal_hoje
                FROM perf_l1_tarefa t
                WHERE t.pessoa_id = ANY(:ids) AND t.status = 'Pendente' AND {_NAO_ACOMP}
                  {_periodo_clause(dias)}
                GROUP BY t.pessoa_id, subtipo
                ORDER BY total DESC
                """
            ),
            {"ids": list(pessoa_ids), "dias": dias},
        ).fetchall()
        return [
            {
                "pessoa_id": r.pessoa_id, "subtipo": r.subtipo, "total": r.total,
                "atrasado": r.atrasado, "fatal_hoje": r.fatal_hoje,
            }
            for r in rows
        ]

    def redistribuir_tarefas(self, team: str, pessoa_id: int, subtipo: str, dias: int, limit: int = 500) -> list[dict]:
        """Tarefas individuais de um (colaborador, subtipo) pro modal de detalhe."""
        sub_clause = "t.subtipo IS NULL" if subtipo == "(sem subtipo)" else "t.subtipo = :sub"
        rows = self.db.execute(
            text(
                f"""
                SELECT t.l1_task_id, t.subtipo, t.cnj, t.pasta, t.uf,
                  t.prazo_previsto, t.cadastrado_em,
                  CASE WHEN t.prazo_previsto IS NULL THEN 'sem_prazo'
                       WHEN {_PRAZO} < {_HOJE} THEN 'atrasado'
                       WHEN {_PRAZO} = {_HOJE} THEN 'fatal_hoje'
                       ELSE 'futuro' END AS situacao
                FROM perf_l1_tarefa t
                WHERE t.pessoa_id = :pid AND t.status = 'Pendente' AND {sub_clause} AND {_NAO_ACOMP}
                  {_periodo_clause(dias)}
                ORDER BY t.prazo_previsto ASC NULLS LAST
                LIMIT :lim
                """
            ),
            {"pid": pessoa_id, "sub": subtipo, "dias": dias, "lim": limit},
        ).fetchall()
        return [
            {
                "l1_task_id": r.l1_task_id, "subtipo": r.subtipo, "cnj": r.cnj,
                "pasta": r.pasta, "uf": r.uf, "situacao": r.situacao,
                "prazo": r.prazo_previsto.isoformat() if r.prazo_previsto else None,
            }
            for r in rows
        ]

    def descricoes(self, ids: list) -> dict:
        """Descrição (assunto/anotações) AO VIVO do L1 pra os task ids dados —
        não vem no snapshot. Batch de 30 (limite do $top em /Tasks)."""
        from app.services.legal_one_client import LegalOneApiClient

        clean = [int(i) for i in ids if i][:150]
        if not clean:
            return {}
        client = LegalOneApiClient()
        out: dict = {}
        for i in range(0, len(clean), 30):
            chunk = clean[i : i + 30]
            flt = "id in (" + ",".join(str(x) for x in chunk) + ")"
            try:
                for t in client.search_tasks(filter_expression=flt, top=30, select="id,description"):
                    out[t.get("id")] = t.get("description")
            except Exception:  # noqa: BLE001 — best-effort; descrição é enriquecimento
                continue
        return out

    # ── Log de redistribuição (aba Relatórios) ──
    def registrar_log(self, team: str, user, movimentos: list) -> dict:
        from app.models.performance import BalanceadorLog

        total_tar = sum(int(m.get("qtd") or 0) for m in (movimentos or []))
        log = BalanceadorLog(
            team=team,
            criado_por_id=getattr(user, "id", None),
            criado_por_nome=getattr(user, "name", None) or getattr(user, "email", None),
            total_movimentos=len(movimentos or []),
            total_tarefas=total_tar,
            origem="mock",
            detalhe=movimentos or [],
        )
        self.db.add(log)
        self.db.commit()
        self.db.refresh(log)
        return {"id": log.id, "total_movimentos": log.total_movimentos, "total_tarefas": log.total_tarefas}

    def listar_logs(self, team: str, limit: int = 50) -> list:
        from app.models.performance import BalanceadorLog

        rows = (
            self.db.query(BalanceadorLog)
            .filter(BalanceadorLog.team == team)
            .order_by(BalanceadorLog.criado_em.desc())
            .limit(limit)
            .all()
        )
        return [
            {
                "id": r.id,
                "criado_em": r.criado_em.isoformat() if r.criado_em else None,
                "criado_por_nome": r.criado_por_nome,
                "total_movimentos": r.total_movimentos,
                "total_tarefas": r.total_tarefas,
                "origem": r.origem,
                "detalhe": r.detalhe or [],
            }
            for r in rows
        ]

    # ── LIVE: pendentes não-iniciadas de uma pessoa, direto do L1 ──
    def live_pessoa(self, team: str, pessoa_id: int, dias: int) -> dict:
        """Pendentes NÃO iniciadas (statusId=0) da pessoa, AO VIVO do L1 (filtro
        por participante). Agrupa por subtipo (nome via catálogo local
        LegalOneTaskSubType) + devolve os detalhes. Base da redistribuição em
        tempo real — o número que o supervisor vê é o de AGORA, não o snapshot."""
        from collections import defaultdict

        from app.models.legal_one import LegalOneTaskSubType
        from app.models.performance import PerfPessoa
        from app.services.legal_one_client import LegalOneApiClient

        p = self.db.query(PerfPessoa).filter(PerfPessoa.id == pessoa_id).first()
        if not p:
            return {"pessoa_id": pessoa_id, "nome": None, "resolvido": False, "subtipos": [], "tarefas": []}
        cid = _user_map().get(p.nome_norm)
        if not cid:
            return {"pessoa_id": pessoa_id, "nome": p.nome, "resolvido": False, "subtipos": [], "tarefas": []}

        client = LegalOneApiClient()
        flt = f"participants/any(pp: pp/contact/id eq {cid}) and statusId eq 0"
        raw, skip = [], 0
        while skip < 900:  # teto de segurança (~30 páginas de 30)
            r = client._request_with_retry(
                "GET",
                f"{client.base_url}/Tasks",
                params={"$filter": flt, "$top": "30", "$skip": str(skip), "$select": "id,subTypeId,deadLine,description"},
            )
            batch = r.json().get("value", [])
            raw.extend(batch)
            if len(batch) < 30:
                break
            skip += 30

        sub_ids = {t.get("subTypeId") for t in raw if t.get("subTypeId")}
        nomes = {
            s.external_id: s.name
            for s in self.db.query(LegalOneTaskSubType).filter(LegalOneTaskSubType.external_id.in_(sub_ids)).all()
        }
        hoje = _hoje_brt()
        teto = hoje + _dt.timedelta(days=dias) if dias and dias > 0 else None

        tarefas = []
        for t in raw:
            sub = nomes.get(t.get("subTypeId")) or f"subtipo {t.get('subTypeId')}"
            if sub.lower().startswith("acompanhar"):
                continue
            dl = t.get("deadLine")
            d = None
            if dl:
                try:
                    d = _dt.datetime.fromisoformat(dl).date()
                except ValueError:
                    d = None
            if d is None:
                sit = "sem_prazo"
            elif d < hoje:
                sit = "atrasado"
            elif d == hoje:
                sit = "fatal_hoje"
            else:
                sit = "futuro"
            if teto is not None and d is not None and d > teto:
                continue  # fora do período
            tarefas.append(
                {
                    "l1_task_id": t.get("id"), "subtipo": sub, "descricao": t.get("description"),
                    "prazo": dl, "situacao": sit,
                }
            )

        agg = defaultdict(lambda: {"total": 0, "atrasado": 0, "fatal_hoje": 0})
        for t in tarefas:
            a = agg[t["subtipo"]]
            a["total"] += 1
            if t["situacao"] == "atrasado":
                a["atrasado"] += 1
            elif t["situacao"] == "fatal_hoje":
                a["fatal_hoje"] += 1
        subtipos = [{"subtipo": k, **v} for k, v in sorted(agg.items(), key=lambda x: -x[1]["total"])]
        return {"pessoa_id": pessoa_id, "nome": p.nome, "resolvido": True, "subtipos": subtipos, "tarefas": tarefas}
